Issue #57 - Allow sorting of changesets by coverage score

armenzg opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/mozilla/firefox-code-coverage-frontend/178)
<!-- Reviewable:end -->

armenzg commented:
This is ready to review. r? @jankeromnes

jankeromnes commented on src/components/summary.jsx:
```diff
@@ -1,24 +1,86 @@
+      sortedChangesets =
+        sortChangesetsByCoverage(changesets, changesetsCoverage, this.state.sortingReversed);
+    }
+    return sortedChangesets;
```
Nit: Same here, you could return early instead of using an intermediate variable.

jankeromnes commented on src/utils/data.js:
```diff
@@ -56,17 +56,45 @@ const sortChangesetsByRecency = (a, b) => {
-  if ((Object.keys(changesets).legnth === 0) ||
-      (Object.keys(changesetsCoverage).length === 0)) {
-    return [];
+const sortWithUndefined = (a, b) => {
```
Nit: Please explain the point of this weird sorting algorithm in a short comment above.

jankeromnes commented on src/utils/data.js:
```diff
@@ -56,17 +56,45 @@ const sortChangesetsByRecency = (a, b) => {
+const sortWithUndefined = (a, b) => {
+  if ((typeof a.percentage === 'undefined') && (typeof b.percentage === 'undefined')) {
+    return 0;
+  } else if (typeof a.percentage === 'undefined') {
```
Nit: No need for `else` after return. Also, if you move this condition to the top, you can simplify the other condition:

```js
if (A && B) {
  return 0;
} else if (A) {
  return 1;
}
return -1;
```

becomes:

```js
if (A) {
  return 1;
}
if (B) {
  return 0;
}
return -1;
```

armenzg replied:
I tried it but it does not work:
![image](https://user-images.githubusercontent.com/44410/40272447-08c06818-5b7b-11e8-9812-9481d78afa22.png)

jankeromnes commented on src/utils/data.js:
```diff
@@ -56,17 +56,45 @@ const sortChangesetsByRecency = (a, b) => {
+const sortChangesetsByCoverageScore = (a, b) => {
+  let retVal;
+  if ((typeof a.percentage === 'undefined') || (typeof b.percentage === 'undefined')) {
+    retVal = sortWithUndefined(a, b);
```
When do we expect an `undefined` percentage? Nit: Please add a short comment.

armenzg replied:
Some changesets are marked as 'No changes'. These changes cannot affect coverage, thus, an undefined percentage.

I could also add an extra property (`noCoverageChanges`) and check for that.
By doing this I would avoid having an implicit meaning for the `percentage` value.

jankeromnes commented on src/utils/data.js:
```diff
@@ -56,17 +56,45 @@ const sortChangesetsByRecency = (a, b) => {
+};
+
+const sortChangesetsByCoverageScore = (a, b) => {
+  let retVal;
```
Nit: No need for an intermediate variable, please return directly.

jankeromnes commented on src/utils/data.js:
```diff
@@ -56,17 +56,45 @@ const sortChangesetsByRecency = (a, b) => {
+    retVal = sortWithUndefined(a, b);
+  } else if (a.percentage < b.percentage) {
+    retVal = -1;
+  } else if (a.percentage === b.percentage) {
```
Nit: You won't need the `else` when you return directly.

jankeromnes commented on src/utils/data.js:
```diff
@@ -56,17 +56,45 @@ const sortChangesetsByRecency = (a, b) => {
+    retVal = -1;
+  } else if (a.percentage === b.percentage) {
+    retVal = 0;
+  } else {
```
Nit: You won't need the `else` when you return directly.

armenzg commented on src/containers/summary.jsx:
```diff
@@ -21,98 +24,58 @@ export default class SummaryContainer extends Component {
-      console.debug(`We have ${changesets.length} changesets.`);
-      console.debug(`pending: ${summary.pending}`);
-      console.debug(`errors: ${summary.error}`);
+      const { changesets, changesetsCoverage, summary } = await loadCoverageData();
```
Better?

```diff
   async componentDidMount() {
-    this.loadCoverageData();
+    this.initializeData();
   }

-  async loadCoverageData() {
+  async initializeData() {
     try {
+      // This will either fetch the data or grab it from the cache
       const { changesets, changesetsCoverage, summary } = await loadCoverageData();
```

armenzg commented on src/containers/summary.jsx:
```diff
@@ -21,98 +24,58 @@ export default class SummaryContainer extends Component {
-      } else {
-        this.setState({ coverage: newCoverage });
-      }
+      const { csetsCoverage, polling } = await pollPendingChangesets(changesetsCoverage);
```
Some of this `changesetsCoverage` vs `coverage` vs `csetsCoverage` is due to the linting rules.

e.g. 
```javascript
const foo = (changesetsCoverage) =>
   const { changesetsCoverage } = bar(changesetsCoverage);
```
becomes of the [no-shadow](https://eslint.org/docs/rules/no-shadow) rule ends up becoming (or `coverage`):
```javascript
const foo = (csetsCoverage) =>
   const { changesetsCoverage } = bar(csetsCoverage);
```
We simply want to take a piece of data and create a new one with some modifications.
It is not pretty and perhaps a change

armenzg commented on src/utils/data.js:
```diff
@@ -6,6 +13,96 @@ export const arrayToMap = (csets = []) => {
+const sortChangesetsByRecency = (a, b) => {
+  if (a.pushId < b.pushId) {
+    return 1;
+  } else if (a.pushId === b.pushId) {
```
That's affected by the [consistent-return](https://eslint.org/docs/rules/consistent-return) rule.

I will use this instead:
```javascript
let retVal;
// no returns used in various if conditions but assigning the return value to retVal
return retVal;

armenzg commented on src/utils/hg.js:
```diff
@@ -88,15 +94,16 @@ const pushesToCsets = async (pushes) => {
-        .forEach((cset) => {
-          filteredCsets.push(initializedChangeset(cset, cset.author));
+        .forEach((cset, changesetIndex) => {
+          filteredCsets[cset.node] =
```
I noticed this issue on another branch I've been working on.
I will cherry-pick it.

armenzg commented on src/utils/hg.js:
```diff
@@ -125,5 +132,3 @@ const getChangesets = async (repoName = REPO_NAME) => {
   return csets;
 };
-
-export default getChangesets;
```
From another module this would work: `import getChangesets from '../utils/hg'`
You could even do `import whateverNameIWant from '../utils/hg';`

However if the function is exported the standard way (e.g. `export const getChangesets = ...`) you would have to import the function like this `import { getChangesets } from '../utils/hg';`

I believe this is all related that a module is exported as an Object with function names as properties, thus, we can use JS destructuring of objects to only obtain the wanted properties.

Does this help? I hope I got it right.

armenzg commented:
Hi @jankeromnes I've addressed all issues. I've postponed one.

Wrt to the if/else approaches, I believe we should just take the current approach or go back to my previous approach (with a return statement at the end). The [consistent return](https://eslint.org/docs/rules/consistent-return) rule kind of gets on the way for other alternatives.

Thanks for your feedback!

Also, please let me know if you would prefer certain kind of PRs and not others (e.g. adding feature versus code clean up).

jankeromnes commented:
Hi @armenzg, as mentioned in https://github.com/mozilla/firefox-code-coverage-frontend/pull/182#issuecomment-391153448, I think your commits got mixed up. Sorry to make you work on multiple branches / pull requests at the same time.

Could you please remove the unrelated refactor commits from this pull request?

armenzg commented:
@jankeromnes I've managed to separate everything out of this PR.

armenzg commented on src/components/summary.jsx:
```diff
@@ -1,24 +1,86 @@
+        sortingMethod: sortingMethods.DATE,
+        sortingReversed: false,
+      });
+    }
```
I think my latest commit addresses this:
https://github.com/mozilla/firefox-code-coverage-frontend/pull/178/commits/b2250f3df0223cee3032a3818cfe21d9e80fedbf

armenzg commented:
In case it got missed. @jankeromnes r?

The deployment is here:
https://firefox-code-coverage-pr-178.herokuapp.com/#/

You can ignore the Travis test failures. That's an orthogonal issue I will be fixing soon.

armenzg commented:
I've rebased this and fixed some coverage test regressions that this PR had introduced.

jankeromnes commented:
Oops, there seem to be many different ways to show that the page is "loading", and some of them seem wrong.

The page first shows this (yellow banner that says polling):

![coverage-loading](https://user-images.githubusercontent.com/599268/40923151-50f18b94-6814-11e8-8e8d-2a53d8fe9b4c.png)

And after a while it changes to this (different loading message, and a weird error that says there is no data, while clearly there actually is data to show):

![coverage-no-data](https://user-images.githubusercontent.com/599268/40923193-6d6eee06-6814-11e8-8e18-3f6a3451762c.png)

jankeromnes commented on src/components/summary.jsx:
```diff
@@ -1,24 +1,76 @@
+            <th>Author</th>
+            <th>Changeset</th>
+            <th>Description</th>
+            <th onClick={() => this.onClickNextSorting()}>
```
Nits:
- It would be great to set the style to `cursor: pointer` to indicate that this can be clicked.
- It would be super helpful to add a `title` attribute (message that display on hover) saying "sorted by date (click to sort by coverage)"

jankeromnes commented:
Does the sorting feature actually work as intended? I get these results on the latest deployment:

- Sorted by (default initial case)?

![coverage-sorted-1](https://user-images.githubusercontent.com/599268/40925662-88cedc82-681a-11e8-94d9-01eb2124a40d.png)

- Click! Sorted by coverage?

![coverage-sorted-2](https://user-images.githubusercontent.com/599268/40925698-9bcf3502-681a-11e8-8759-2dbe4bd5047b.png)

- Click! Sorted by reverse coverage?

![coverage-sorted-3](https://user-images.githubusercontent.com/599268/40925738-ab46b9b0-681a-11e8-8a23-b0d58e3e9175.png)

jankeromnes commented:
> armenzg deployed to [firefox-code-coverage-pr-178](https://firefox-code-coverage-pr-178.herokuapp.com/) 22 hours ago

Whoa, works much better now! Thanks :smile: 

I'll make another quick pass on the code, but the functionality is there and actually works very nice (I wasn't sold on the three state sorting, but now it's totally obvious how it works, great job!)

jankeromnes commented on README.md:
````diff
@@ -54,3 +54,6 @@ Checkout the code and run:
 ```
+
+# Attribution
+Icons from [Material Design Icons](https://material.io/tools/icons/?style=baseline).
````
Feels like I've reviewed this line 10 times now :stuck_out_tongue_closed_eyes: Approved!

jankeromnes commented on src/components/summary.jsx:
```diff
@@ -1,24 +1,84 @@
+            <th>Description</th>
+            <th
+              onClick={() => this.onClickNextSorting()}
+              title={'sorted by date (click to sort by coverage)'}
```
Nit: This title attribute doesn't change, so it's only correct in one of the three sorting states. But since the UX is now fairly obvious, I feel like maybe we don't even need a title. Please feel free to fix it or to remove it, as you prefer.

To fix it, I'd suggest cycling through the following values:
- `sorted by date (click to sort by coverage)`
- `sorted by coverage, lowest first (click to reverse)`
- `sorted by coverage, highest first (click to sort by date)`
